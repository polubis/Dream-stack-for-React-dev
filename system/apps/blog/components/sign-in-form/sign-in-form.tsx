import { Alert, Box, Button, Font, Input } from '@system/figa-ui';
import { useSignInStore } from '../../store/sign-in';
import { useSignOutStore } from '../../store/sign-out';
import { useState } from 'react';
import { get } from '@system/blog-selectors';

const SignInForm = () => {
  const signInStore = useSignInStore();
  const signOutStore = useSignOutStore();
  const [login, setLogin] = useState('');
  const [password, setPassword] = useState('');

  const handleSignIn = () => {
    signInStore.signIn({
      login,
      password,
    });
  };

  return (
    <>
      <Box spacing={[400, 150, 400, 400]} maxWidth="320px" margin="auto">
        <Font variant="h6">Sign in into your account</Font>
        <Input
          value={login}
          autoFocus
          data-i={get('sign-in-login-input')}
          placeholder="Login*"
          onChange={(e) => setLogin(e.target.value)}
        />
        <Input
          value={password}
          type="password"
          data-i={get('sign-in-password-input')}
          placeholder="Password*"
          onChange={(e) => setPassword(e.target.value)}
        />
        <Button
          data-i={get('sign-in-confirm-btn')}
          loading={
            signInStore.key === 'pending' || signOutStore.key === 'pending'
          }
          onClick={handleSignIn}
        >
          Confirm
        </Button>
        {signInStore.key === 'error' && (
          <Alert data-i={get('sign-in-error-alert')} type="error">
            {signInStore.response.message}
          </Alert>
        )}
        {signInStore.key === 'ok' && (
          <Alert data-i={get('sign-in-ok-alert')} type="ok">
            You are logged in!
          </Alert>
        )}
      </Box>
    </>
  );
};

export { SignInForm };
