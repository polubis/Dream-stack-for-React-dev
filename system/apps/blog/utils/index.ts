import type { Lang } from '@system/blog-api-models';
import { readdir } from 'fs';
import path from 'path';

export const createArticlePath = (lang: Lang, ...rest: string[]): string =>
  path.join(
    process.cwd(),
    'apps',
    'blog',
    'content',
    lang,
    'articles',
    ...rest
  );

export const getArticlesIds = (lang: Lang): Promise<string[]> => {
  return new Promise((resolve, reject) => {
    const dirPath = createArticlePath(lang);

    readdir(dirPath, { withFileTypes: true }, (err, files) => {
      if (err) {
        reject(err);
      }

      resolve(
        files
          .filter((file) => file.isFile())
          .map((file) => file.name.replace('.mdx', ''))
      );
    });
  });
};
