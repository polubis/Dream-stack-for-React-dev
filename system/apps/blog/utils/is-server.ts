const isServer = (): boolean => typeof window === 'undefined';

const isClient = (): boolean => !isServer();

export { isServer, isClient };
