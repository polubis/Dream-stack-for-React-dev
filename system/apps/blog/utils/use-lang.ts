import { useRouter } from 'next/router';
import { Lang } from '../models';
import { useMemo } from 'react';

const isLang = (param: string): param is Lang => {
  return param === ('en' as Lang) || param === ('pl' as Lang);
};

const getLang = (pathname: string): Lang => {
  const lang = pathname.split('/')[1];

  return isLang(lang) ? lang : 'en';
};

const useLang = () => {
  const router = useRouter();

  return useMemo(() => getLang(router.pathname), [router]);
};

export { useLang };
