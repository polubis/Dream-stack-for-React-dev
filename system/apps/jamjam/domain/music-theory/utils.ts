import type { NoteId, NoteNotation, NoteNotationSymbol } from './defs';
import {
  BMOLL_NOTE_NOTATION_SYMBOLS,
  SHARP_NOTE_NOTATION_SYMBOLS,
} from './consts';

export const getNoteSymbol = (
  noteId: NoteId,
  notation: NoteNotation
): NoteNotationSymbol => {
  const id = noteId - 1;

  if (notation === 'sharp') {
    return SHARP_NOTE_NOTATION_SYMBOLS[id];
  }

  return BMOLL_NOTE_NOTATION_SYMBOLS[id];
};
