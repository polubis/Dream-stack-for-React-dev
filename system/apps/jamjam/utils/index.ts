const getLast = <T>(items: T[]): T | undefined =>
  (Array.isArray(items) ? items : []).at(-1);

export { getLast };
