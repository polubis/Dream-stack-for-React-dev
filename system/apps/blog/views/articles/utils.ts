import {
  ArticleStatus,
  GetArticlesParams,
  Lang,
} from '@system/blog-api-models';
import { isServer } from '@system/utils';
import { useSearchParams } from 'next/navigation';

const getSearchParam = (param: unknown, defaultSearch: string): string => {
  return typeof param === 'string' ? param : defaultSearch;
};

const getNumericParam = (param: unknown, defaultValue: number): number => {
  if (typeof param !== 'string') return defaultValue;
  const parsedParam = Number.parseInt(param);
  return Number.isNaN(parsedParam) ? defaultValue : parsedParam;
};

const isArticleStatus = (status: unknown): status is ArticleStatus => {
  return (
    ['Accepted', 'Draft', 'NeedWork', 'WaitingForApproval'] as ArticleStatus[]
  ).includes(status as ArticleStatus);
};

const getStatusParam = (
  param: unknown,
  defaultStatus: ArticleStatus
): ArticleStatus => {
  if (!isArticleStatus(param)) return defaultStatus;
  return param;
};

const getSearchParams = (
  lang: Lang,
  searchParams: ReturnType<typeof useSearchParams>
): Readonly<GetArticlesParams> => {
  const defaults: GetArticlesParams = {
    Search: '',
    ItemsPerPage: 20,
    CurrentPage: 1,
    Status: 'Accepted',
    lang,
  };

  if (isServer()) {
    return defaults;
  }

  const params: GetArticlesParams = {
    Search: getSearchParam(searchParams.get('Search'), defaults.Search),
    ItemsPerPage: getNumericParam(
      searchParams.get('ItemsPerPage'),
      defaults.ItemsPerPage
    ),
    CurrentPage: getNumericParam(
      searchParams.get('CurrentPage'),
      defaults.CurrentPage
    ),
    Status: getStatusParam(searchParams.get('Status'), defaults.Status),
    lang,
  };

  return params;
};

export { getSearchParams };
