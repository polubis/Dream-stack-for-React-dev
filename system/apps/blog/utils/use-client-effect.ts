import { useEffect } from 'react';
import { isClient } from './is-server';

type CleanUp = void | (() => void);

const useClientEffect = (effect: () => CleanUp, deps: unknown[] = []): void => {
  useEffect(() => {
    if (isClient()) {
      return;
    }

    const cleanUp = effect();

    return () => {
      cleanUp && cleanUp();
    };
  }, deps);
};

export { useClientEffect };
