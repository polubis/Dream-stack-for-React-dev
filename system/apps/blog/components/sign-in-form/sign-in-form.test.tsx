import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { SignInForm } from './sign-in-form';
import {
  BlogEnv,
  getPath,
  mockErrorResponse,
  requestFixture,
} from '@system/blog-api';
import { storage } from '@system/utils';

describe('User can use form to sign in when: ', () => {
  const authStorage = storage<BlogEnv>();
  const { clean, server, rest } = requestFixture();

  beforeAll(() => {
    server.listen();
  });

  afterEach(() => {
    clean();
    authStorage.clear();
  });

  afterAll(() => {
    server.close();
  });

  const LOGIN = 'tom19944';
  const PASSWORD = 'my-password-1994';

  const signInFormTest = () => {
    const result = render(<SignInForm />);

    expect(screen.getByText(/Confirm/).closest('button').disabled).toBeFalsy();

    fireEvent.change(screen.getByPlaceholderText('Login*'), {
      target: { value: LOGIN },
    });
    fireEvent.change(screen.getByPlaceholderText('Password*'), {
      target: { value: PASSWORD },
    });
    fireEvent.click(screen.getByText(/Confirm/));

    screen.getByDisplayValue(LOGIN);
    screen.getByDisplayValue(PASSWORD);
    expect(screen.getByText(/Confirm/).closest('button').disabled).toBeTruthy();

    return result;
  };

  it('goes through success flow', async () => {
    const { asFragment } = signInFormTest();

    await waitFor(() => {
      screen.getByText(/You are logged in!/);
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('goes through failure flow', async () => {
    server.use(
      rest.post(getPath('Account/SignIn'), (_, res, ctx) => {
        return res(ctx.status(404), ctx.json(mockErrorResponse()));
      })
    );

    const { asFragment } = signInFormTest();

    await waitFor(() => {
      screen.getByText(mockErrorResponse().errors[0].message);
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
