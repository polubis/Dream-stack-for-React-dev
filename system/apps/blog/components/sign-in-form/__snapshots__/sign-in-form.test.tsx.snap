// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User can use form to sign in when:  goes through failure flow 1`] = `
<DocumentFragment>
  <div
    class="box column empty"
    style="margin: auto; max-width: 320px;"
  >
    <div
      class="box-item-wrapper"
      style="margin-bottom: 32px;"
    >
      <h6
        class="font h6 default"
      >
        Sign in into your account
      </h6>
    </div>
    <div
      class="box-item-wrapper"
      style="margin-bottom: 12px;"
    >
      <div
        class="input"
      >
        <input
          data-i="sign-in-login-input"
          placeholder="Login*"
          value="tom19944"
        />
      </div>
    </div>
    <div
      class="box-item-wrapper"
      style="margin-bottom: 32px;"
    >
      <div
        class="input"
      >
        <input
          data-i="sign-in-password-input"
          placeholder="Password*"
          type="password"
          value="my-password-1994"
        />
      </div>
    </div>
    <div
      class="box-item-wrapper"
      style="margin-bottom: 32px;"
    >
      <button
        class="button size-3 rectangle filled primary"
        data-i="sign-in-confirm-btn"
      >
        Confirm
      </button>
    </div>
    <div
      class="box-item-wrapper"
    >
      <div
        class="alert error filled"
        data-i="sign-in-error-alert"
      >
        <p
          class="font b2 default"
        >
          Something went wrong...
        </p>
        <svg
          class="icon error-icon"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1.5 12C1.5 14.0767 2.11581 16.1068 3.26957 17.8335C4.42332 19.5602 6.0632 20.906 7.98182 21.7007C9.90045 22.4955 12.0116 22.7034 14.0484 22.2982C16.0852 21.8931 17.9562 20.8931 19.4246 19.4246C20.8931 17.9562 21.8931 16.0852 22.2982 14.0484C22.7034 12.0116 22.4955 9.90045 21.7007 7.98182C20.906 6.0632 19.5602 4.42332 17.8335 3.26957C16.1068 2.11581 14.0767 1.5 12 1.5C9.21523 1.5 6.54451 2.60625 4.57538 4.57538C2.60625 6.54451 1.5 9.21523 1.5 12ZM18.8625 17.8125L6.1875 5.1375C7.91298 3.70125 10.1129 2.96173 12.3556 3.0641C14.5983 3.16646 16.7218 4.10331 18.3092 5.69077C19.8967 7.27824 20.8335 9.40169 20.9359 11.6444C21.0383 13.8871 20.2988 16.087 18.8625 17.8125ZM6.18 18.87C4.36374 17.3276 3.23325 15.1279 3.03642 12.7533C2.83959 10.3787 3.59248 8.02286 5.13 6.2025L17.7975 18.87C16.1732 20.2425 14.1153 20.9955 11.9887 20.9955C9.86219 20.9955 7.80433 20.2425 6.18 18.87Z"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`User can use form to sign in when:  goes through success flow 1`] = `
<DocumentFragment>
  <div
    class="box column empty"
    style="margin: auto; max-width: 320px;"
  >
    <div
      class="box-item-wrapper"
      style="margin-bottom: 32px;"
    >
      <h6
        class="font h6 default"
      >
        Sign in into your account
      </h6>
    </div>
    <div
      class="box-item-wrapper"
      style="margin-bottom: 12px;"
    >
      <div
        class="input"
      >
        <input
          data-i="sign-in-login-input"
          placeholder="Login*"
          value="tom19944"
        />
      </div>
    </div>
    <div
      class="box-item-wrapper"
      style="margin-bottom: 32px;"
    >
      <div
        class="input"
      >
        <input
          data-i="sign-in-password-input"
          placeholder="Password*"
          type="password"
          value="my-password-1994"
        />
      </div>
    </div>
    <div
      class="box-item-wrapper"
      style="margin-bottom: 32px;"
    >
      <button
        class="button size-3 rectangle filled primary"
        data-i="sign-in-confirm-btn"
      >
        Confirm
      </button>
    </div>
    <div
      class="box-item-wrapper"
    >
      <div
        class="alert ok filled"
        data-i="sign-in-ok-alert"
      >
        <p
          class="font b2 default"
        >
          You are logged in!
        </p>
        <svg
          class="icon success-icon"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10.243 16.3139L6 12.0699L7.414 10.6559L10.243 13.4839L15.899 7.8269L17.314 9.2419L10.243 16.3119V16.3139Z"
          />
          <path
            clip-rule="evenodd"
            d="M1 12C1 5.925 5.925 1 12 1C18.075 1 23 5.925 23 12C23 18.075 18.075 23 12 23C5.925 23 1 18.075 1 12ZM12 21C10.8181 21 9.64778 20.7672 8.55585 20.3149C7.46392 19.8626 6.47177 19.1997 5.63604 18.364C4.80031 17.5282 4.13738 16.5361 3.68508 15.4442C3.23279 14.3522 3 13.1819 3 12C3 10.8181 3.23279 9.64778 3.68508 8.55585C4.13738 7.46392 4.80031 6.47177 5.63604 5.63604C6.47177 4.80031 7.46392 4.13738 8.55585 3.68508C9.64778 3.23279 10.8181 3 12 3C14.3869 3 16.6761 3.94821 18.364 5.63604C20.0518 7.32387 21 9.61305 21 12C21 14.3869 20.0518 16.6761 18.364 18.364C16.6761 20.0518 14.3869 21 12 21Z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
